Seq.delay : (unit -> 'a t) -> 'a t

gasche opened the pull request:
Quoting the new docstring:

```ocaml
    [delay (fun () -> seq)] has the same elements as [seq], but the expression
    [seq] is only evaluated when the first element is requested.

   For example,
   {[
   let rec of_tree t =
     match t with
     | Leaf v -> Seq.return v
     | Node(l, r) -> Seq.append (of_tree l) (of_tree r)
   ]}
   does not behave as expected, as it traverses its input tree strictly
   even if the output sequence is not forced.
   This function could instead be written as follows:
   {[
   let rec of_tree t =
     Seq.delay @@ fun () ->
     match t with
     | Leaf v -> Seq.return v
     | Node(l, r) -> Seq.append (of_tree l) (of_tree r)
   ]}
```

**Edit**: a previous iteration of this PR also proposed `flatten`=>`concat` functions, now moved to a separate PR #10352.

gasche committed:
Seq.delay : (unit -> 'a t) -> 'a t

dbuenzli commented:
[`concat`](https://github.com/ocaml/ocaml/pull/8760#issuecomment-515059721) ?

ulugbekna commented on stdlib/seq.ml:
```diff
@@ -50,16 +52,24 @@ let rec filter f seq () = match seq() with
   | Nil -> Nil
   | Cons (x, next) ->
-    flat_map_app f (f x) next ()
+     append_flatten x next ()
+(* [append_flatten seq tail] is [append seq (flatten tail)] *)
+and append_flatten seq tail () = match seq () with
+  | Nil -> flatten tail ()
+  | Cons (x, next) -> Cons (x, append_flatten next tail)
 
-(* this is [append seq (flat_map f tail)] *)
-and flat_map_app f seq tail () = match seq () with
+let rec flat_map f seq () = match seq () with
+  | Nil -> Nil
+  | Cons (x, next) ->
+    append_flat_map (f x) f next ()
+(* [append_flat_map seq f tail] is [append seq (flat_map f tail)] *)
+and append_flat_map seq f tail () = match seq () with
   | Nil -> flat_map f tail ()
   | Cons (x, next) ->
-    Cons (x, flat_map_app f next tail)
+    Cons (x, append_flat_map next f tail)
```
What's the benefit of auxiliary mutually recursive functions for `flatten` and `flat_map` over simpler implementations below? 

```
  let rec concat (s1 : 'a t) (s2 : 'a t) : 'a t =
   fun () ->
    match s1 () with Nil -> s2 () | Cons (v, rest) -> Cons (v, concat rest s2)

  let rec flatten (ss : 'a t t) : 'a t =
   fun () ->
    match ss () with
    | Nil -> Nil
    | Cons (v, rest) -> (concat v (flatten rest)) ()

  let rec flat_map (f : 'a -> 'b t) (s : 'a t) : 'b t =
   fun () ->
    match s () with
    | Nil -> Nil
    | Cons (v, rest) -> concat (f v) (flat_map f rest) ()
```

`concat` would be `Seq.append`.

gasche replied:
I don't have a very convincing answer to this question, so I think I will just rewrite the code as you suggest.

Somewhat-explanatory remarks:
- The code of flat_map was written in this style originally by @c-cube, so it felt natural to reuse the structure
- When writing standard-library code, it is somewhat common to write in a fairly low-level style with some parts manually inlined (but here I could not convince myself that it might have any performance benefit).
- In strict versions of the same functions (for lists typically), the inlining has the benefit that it exposes an opportunity for TRMC transformation (see #9760). Here this aspect is irrelevant as recursive calls return a function closure immediately, so the call-stack usage remains constant.

ulugbekna replied:
Thanks for remarks! It's a pity that trmc PR didn't get merged

gasche replied:
I changed the implementation as suggested. Thanks!

ulugbekna reviewed (commented):
Interestingly, I was working through parser combinators exercise by F. Pottier and was surprised `Stdlib.Seq` didn't contain the functions implemented in this PR and `Seq.hd/take`.

gasche commented:
I renamed `flatten` into `concat` and added `concat_map` as an alias for `flat_map`.

(For the record, I find `flatten` more descriptive than `concat`, which is mostly a synonym for `append` to me and does not imply nested structure. But `concat_map` reads better than `flatten_map` and `flat_map` is not standard, so oh well.)

gasche commented:
So are we interested in the new functions proposed in this PR? Cc @nojb the indefatigable stdlib curator, @yallop who made good-taste suggestions for Seq extensions, and @c-cube the original author.

yallop commented:
`concat` is a very useful addition and `flat_map` is a fine alias.  I think we should add them both.

I have some reservations about `delay`, perhaps mostly around the name, since it seems to make its argument less delayed rather than more.

gasche commented:
The way I think of it, `Seq.delay (fun () -> foo)` delays the computation of `foo`. If `foo` is already a value, then it is neither more nor less delayed.

I observed its usefulness by seeing sequence beginners writing "too eager" definitions as in the first `of_tree` version above.

gasche commented:
I proposed `Seq.concat` (and `Seq.concat_map`) in #10352 and will restrict the present PR to `Seq.delay` only.

xavierleroy commented:
@fpottier : what do you think about this proposal, in light of your recent extensions to module Seq ?

fpottier commented:
Indeed, I have felt the need for this `delay` function when writing programming exercises involving the "non-determinism monad".

The need for `delay` is not widely recognized, I think, because most of the folklore on monads has been developed in the Haskell world, where laziness is everywhere by default, so there is no need for `delay`. In a strict language like OCaml, there can be a need for it.

xavierleroy commented:
Thanks @fpottier.  Could you and @gasche discuss the matter urgently, so that we know whether to rush `Seq.delay` in 4.14 ?

gasche commented:
Honestly I don't think that it's worth rushing this function in 4.14. we have done without for years, we can do without for a bit more. I would of course be happy if people came to a decision on it (accept or reject), and I guess I should have been more pushy about it (I usually tend to err on the too-pushy side). But I would prefer this discussion to happen in more relaxed conditions than "last days remaining before the release!" -- time to bike-shed the name, etc.

gasche commented:
This being said: obviously I agree that this function is useful and I'm happy to rebase the PR if there is consensus in favor of having it. (I would expect such a consensus to come from, for example, @yallop, @fpottier, @c-cube, @dbuenzli and @CraigFe.)

gasche commented:
Note: `Seq.delay (fun () -> seq)` can also be written `(fun () -> seq ())`; the problem with this desugared formulation are that:
1. the intent is less clear, and in particular
2. it is harder for beginners to guess (they may not think that they need this unless they are told about `delay` and why it is useful), and finally
3. this "breaks the abstraction" of the `'a Seq.t` type, as it requires knowing that it is of the form `unit -> ...`; it would need to be written differently with another definition of sequences, whereas a `delay` combinator at this type always exists.

dbuenzli commented:
As I said in the `Seq` big bang PR I'm not a heavy `Seq` users so I don't have that much opinion on the matter.  But the name and sig looks good to me. 

I was just wondering why this wasn't maybe `Seq.delay : 'a t Lazy.t -> 'a t`

c-cube commented:
>     3. this "breaks the abstraction" of the `'a Seq.t` type, as it requires knowing that it is of the form `unit -> ...`; it would need to be written differently with another definition of sequences, whereas a `delay` combinator at this type always exists.

There is no abstraction in Seq already, why does it matter?

gasche commented:
@dbuenzli `unit -> ...` is conceptually simpler, what Seq is already using internally (this is not a lazy/memoized list but a thunked list), and I suspect that it produces simpler / more efficient code in the common usage case where a function literal is passed.

@c-cube While the definition is public, users can do most of their programming without needing to know what the definition is, which means that they are likely to forget the definition. Programming is simpler for them if they can consider it abstract.

gasche commented:
I would have had a use of this function today, to build a singleton sequence from a single value, where the value should only be computed when the element of the sequence is forced/demanded.

nojb commented:
Going over the discussion there seems to be a a soft consensus for this function. @gasche: do you want to rebase it and we can try to have it merged?

gasche commented:
Thanks! I rebased the PR.

nojb commented on stdlib/seq.mli:
```diff
@@ -434,6 +434,32 @@ val iterate : ('a -> 'a) -> 'a -> 'a t
+     | Node(l, r) -> Seq.append (of_tree l) (of_tree r)
+   ]}
+
+   @since 4.13
```
```suggestion
   @since 5.4
```

dbuenzli replied:
5.5 or 5.6 ?

nojb replied:
Oops, sorry, I must be tired. The 5.5 branch has not yet been cut so for a small PR such as this one, I think we can target 5.5 (assuming we get the second approval in time).

gasche replied:
Ha! Thanks both for your vigilance. I put 5.5 and am happy to change again if necessary.

nojb reviewed (approved):
LGTM

A second official approval is needed. Perhaps @fpottier?

fpottier commented:
Looks good to me. (I haven't tried to understand the change in `parmatch.ml`, as the code seems a bit tricky.)

nojb commented:
> Looks good to me. (I haven't tried to understand the change in `parmatch.ml`, as the code seems a bit tricky.)

Thanks for the approval. Regarding `parmatch.ml`: I checked that the new code is equivalent to the old code. The old code was particularly unclear; the new code is still a bit hard to decipher but a clear improvement.
